### PR TITLE
Add filtered field to packet source stats, add telemetry framework support

### PR DESCRIPTION
**Pull request by timwoj**
This PR does two things:

- Adds a new field to the statistics that packet sources can return for reporting filtered packets. I sadly have no way to test this because libpcap doesn't return that kind of information. Packet sources that do provide that information can update their `Statistics()` method to return the data.
- Adds method to `PktSrc` for posting statistics data to the telemetry framework. By default this is being called from the `get_net_stats` BIF that gets called every 5 minutes by the `net_stats_update` event (amongst a few other places).

It also includes a little bit of cleanup to that BIF, which looked like leftovers from when we supported multiple packet sources at once.

Sample of the data reported to telemetry, running on a machine here with really low traffic:

```
# HELP zeek_packet_statistics Packet Statistics
# TYPE zeek_packet_statistics counter
zeek_packet_statistics{endpoint="",value="received"} 3654 1680544210254
zeek_packet_statistics{endpoint="",value="dropped"} 0 1680544210254
zeek_packet_statistics{endpoint="",value="link"} 3660 1680544210254
zeek_packet_statistics{endpoint="",value="bytes"} 908323 1680544210254
zeek_packet_statistics{endpoint="",value="filtered"} 0 1680544210254
```

### CODE REVIEW

File: src/RunState.cc
Comment (awelzel) [inline src/RunState.cc:438]: I probably would've tried to keep the single `reporter->Info()` invocation with a possibly empty filtered_msg suffix formatted in. Should be less duplicated and less lines.
```
std::string filtered_msg = "";
if ( s.filtered )
   filtered_msg = zeek::util::fmt(" ...filtered")

reporter->Info(... %s", ..., filtered_msg.c_str());
```

File: src/RunState.cc
Comment (awelzel) [inline src/RunState.cc:432]:
```suggestion
			if ( s.filtered.value() > 0 )
```

Also wonder if a small lambda `pct = [](uint64_t v, uint64_t received) { ... }` would make it a bit shorter.

**Review by awelzel — CHANGES_REQUESTED** [review]

File: src/stats.bif
Comment (timwoj) [reply src/stats.bif:47]: > My take (currently) is that we either use Inc() directly on metrics during processing in C++. 

Part of the problem is that `PktSrc::Statistics()` is a pure-virtual method that has to be implemented by every packet source, and as such we can't count on maintainers to always update the telemetry. For the pcap source, for example, we don't have the data on packet drop counts unless Statistics() is called because that's where `pcap_stats` is called. We can obviously update `received` and `bytes_received` based on the packet data in other places, but now we're doing it in multiple locations (and expecting packet source maintainers to know to do the same). Doing it from a single location takes it out of the maintainers hands of having to deal with.

Comment (timwoj) [reply src/stats.bif:47]: > The following is what's used in zeek-demo to produce 3 of the 5 metrics listed here using the Telemetry::sync hook that runs every (for Prometheus a more reasonable) 10 seconds by default

Why is this part of some demo code and not actually in the zeek code base? That seems rather useful.

Comment (timwoj) [reply src/stats.bif:47]: Ok, looking into it more that's definitely less effort. I was unaware that all of that sample code existed for the telemetry stuff. Odd that it fit perfectly for the github issue.

File: src/iosource/PktSrc.cc
Comment (timwoj) [reply src/iosource/PktSrc.cc:32]: Modified the field names to look like:

```
# HELP zeek_net_stats_link_packets_total Total number of packets on the packet source link before filtering
# TYPE zeek_net_stats_link_packets_total counter
zeek_net_stats_link_packets_total{endpoint=""} 310.000000 1680651547707
# HELP zeek_net_stats_packets_dropped_total Total number of packets dropped
# TYPE zeek_net_stats_packets_dropped_total counter
zeek_net_stats_packets_dropped_total{endpoint=""} 0.000000 1680651547707
# HELP zeek_net_stats_packets_received_total Total number of packets received
# TYPE zeek_net_stats_packets_received_total counter
zeek_net_stats_packets_received_total{endpoint=""} 310.000000 1680651547707
# HELP zeek_net_stats_bytes_received_total Total number of bytes received
# TYPE zeek_net_stats_bytes_received_total counter
zeek_net_stats_bytes_received_total{endpoint=""} 86839.000000 1680651547707
```

File: scripts/policy/misc/stats.zeek
Comment (awelzel) [inline scripts/policy/misc/stats.zeek:93]:
```suggestion
    $name="received-bytes",
```

File: scripts/policy/misc/stats.zeek
Comment (awelzel) [inline scripts/policy/misc/stats.zeek:100]:
```suggestion
    $name="received-packets",
```

File: scripts/policy/misc/stats.zeek
Comment (awelzel) [inline scripts/policy/misc/stats.zeek:107]:
```suggestion
    $name="dropped-packets",
```

File: scripts/policy/misc/stats.zeek
Comment (awelzel) [inline scripts/policy/misc/stats.zeek:114]:
```suggestion
    $name="link-packets",
```

File: scripts/policy/misc/stats.zeek
Comment (awelzel) [inline scripts/policy/misc/stats.zeek:121]:
```suggestion
    $name="filtered-packets",
```

File: scripts/policy/misc/stats.zeek
Comment (awelzel) [inline scripts/policy/misc/stats.zeek:125]: I'm now wondering if a `gauge` `zeek_net_packet_lag_seconds` would be cool, too :-) But can do as a follow-up or when it comes up.

File: scripts/policy/misc/stats.zeek
Comment (awelzel) [reply scripts/policy/misc/stats.zeek:92]: Also going to suggest flip description and unit below for all of them, the zeek-demo examples got that wrong and also put dashes to make Dominik happy.

File: src/RunState.cc
Comment (awelzel) [inline src/RunState.cc:435]:
```suggestion
			double filtered_pct = s.filtered.value() > 0 pct(s.filtered.value(), s.received) : 0.0;
```

Otherwise this may trigger 0 division if a packet source fills it but no packets were received at all.

File: src/iosource/PktSrc.h
Comment (awelzel) [inline src/iosource/PktSrc.h:396]: This can be removed after removing `PostTelemetry()` into script-land?

File: src/iosource/PktSrc.cc
Comment (awelzel) [reply src/iosource/PktSrc.cc:32]: Can now remove the family here.

**Review by awelzel — COMMENTED** [review]
I'd approve, but had too many things to comment on.

Metrics exposed is almost an API we offer that we can't change freely once released, so would rather we spend a bit more time on the naming etc.

Also cc'ing @JustinAzoff . He's probably seen 3 or 4 variations of those metrics in Prometheus now, so may have more thoughts :-)

File: src/stats.bif
Comment (awelzel) [reply src/stats.bif]: > Odd that it fit perfectly for the github issue.

Well, there have been various versions of these metrics in various places, so you'd think at some point they work out mostly :-)

File: scripts/policy/misc/stats.zeek
Comment (awelzel) [inline scripts/policy/misc/stats.zeek:202]:
```suggestion
			info$pkts_filtered = ns$pkts_filtered - last_ns$pkts_filtered;
```

File: scripts/policy/misc/stats.zeek
Comment (timwoj) [reply scripts/policy/misc/stats.zeek:92]: I dunno where I got the `.` from. Took that out.

I'm not sure the point of changing the names to use dashes if CAF is just going to normalize them to underscores anyways.

Comment (awelzel) [reply scripts/policy/misc/stats.zeek:92]: The motivation for the `-` has been that those metrics may be seen elsewhere (JSON API, log files) and existing metrics had used `-`. But yes, if we promote and are concerned mostly with Prometheus, it's very unnatural to use dash and I've over and over just used `_` anyway and for the Prometheus use-case it doesn't make a difference.

Comment (awelzel) [reply scripts/policy/misc/stats.zeek:92]: > I dunno where I got the . from. Took that out.

The latest version still has it (and stats included and the unit isn't at the end). Assuming you have these changes locally stashed?
```
    $prefix="zeek.net_stats",
    $name="packets_filtered",
```

Comment (timwoj) [reply scripts/policy/misc/stats.zeek:92]: Yeah, I hadn't pushed those because we were still talking about the dashes.

**Review by awelzel — APPROVED** [review]
I'm leaving an approval. There's still a compilation issue and probably the baselines need figuring out, but :+1: for where it is.

**Pull request merged**
